lib/remap-functions/src/parse_key_value.rs

```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+/// An undelimited value is all the text until our field_delimiter, or if it is the last value in the line,
+/// just take the rest of the string.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
+fn parse_undelimited<'a>(field_delimiter: &'a str) -> impl Fn(&'a str) -> IResult<&str, &str> {
+    move |input| map(alt((take_until(field_delimiter), rest)), |s: &str| s.trim())(input)
+}
+
+/// Parses the value.
+/// The value has two parsing strategies.
+///
+/// 1. Parse as a delimited field - currently the delimiter is hardcoded to a `"`.
+/// 2. If it does not start with one of the trim values, it is not a delimited field and we parse up to
+///    the next field_delimiter or the eof.
+///
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
+fn parse_value<'a>(field_delimiter: &'a str) -> impl Fn(&'a str) -> IResult<&str, Value> {
+    move |input| {
+        map(
+            alt((parse_delimited('"'), parse_undelimited(field_delimiter))),
+            Into::into,
+        )(input)
+    }
+}
+
+/// Parses the key.
+/// Parsing strategies are the same as parse_value, but we don't need to convert the result to a `Value`.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
+fn parse_key<'a>(field_delimiter: &'a str) -> impl Fn(&'a str) -> IResult<&str, &str> {
+    move |input| alt((parse_delimited('"'), parse_undelimited(field_delimiter)))(input)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use remap::value;
+    use value::Kind;
+
+    #[test]
+    fn test_parse() {
+        assert_eq!(
+            Ok(("", ("ook".to_string(), "pook".into()))),
+            parse_key_value("=", " ")("ook=pook")
+        );
+    }
+
+    #[test]
+    fn test_parse_line() {
+        assert_eq!(
+            Ok((
+                "",
+                vec![
+                    ("ook".to_string(), "pook".into()),
+                    (
+                        "@timestamp".to_string(),
+                        "2020-12-31T12:43:22.2322232Z".into()
+                    ),
+                    ("key#hash".to_string(), "value".into()),
+                    ("key=with=special=characters".to_string(), "value".into()),
+                    ("key".to_string(), "with special=characters".into()),
+                ]
+            )),
+            parse_line(
+                r#"ook=pook @timestamp=2020-12-31T12:43:22.2322232Z key#hash=value "key=with=special=characters"=value key="with special=characters""#,
+                "=",
+                " "
+            )
+        );
+    }
+
+    #[test]
+    fn test_parse_value() {
+        // delimited
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")(r#""noog""#));
+
+        // undelimited
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")("noog"));
+    }
+
+    test_type_def![
+        value_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from("foo").boxed(),
```

Comment:
Nit: it doesn't matter too much, but for future reference, we have a `lit!` macro to save a few keystrokes, since we're using this often in tests (you still need to call `boxed` though).